File: eMRTDReader/sdk/src/main/java/com/example/emrtdreader/analyzer/MrzImageAnalyzer.java
package com.example.emrtdreader.analyzer;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.Rect;
import android.graphics.YuvImage;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;

import com.example.emrtdreader.models.MrzResult;
import com.example.emrtdreader.models.OcrResult;
import com.example.emrtdreader.ocr.DualOcrRunner;
import com.example.emrtdreader.ocr.MrzAutoDetector;
import com.example.emrtdreader.ocr.OcrEngine;
import com.example.emrtdreader.ocr.RectAverager;
import com.example.emrtdreader.utils.MrzBurstAggregator;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * CameraX analyzer that:
 * - finds MRZ band automatically (deterministic)
 * - stabilizes ROI across frames
 * - runs OCR (MLKit/Tesseract or dual)
 * - aggregates MRZ across bursts
 */
public class MrzImageAnalyzer implements ImageAnalysis.Analyzer {

    public interface Listener {
        void onOcr(OcrResult ocr, MrzResult bestSingle, Rect roi);
        void onFinalMrz(MrzResult finalMrz, Rect roi);
    }

    private final Context appContext;
    private final Listener listener;
    private final MrzBurstAggregator aggregator;
    private final RectAverager rectAverager;
    private final AtomicBoolean finished = new AtomicBoolean(false);

    private volatile OcrEngine mlKitEngine;
    private volatile OcrEngine tessEngine;
    private volatile DualOcrRunner.Mode mode;

    private long lastTs = 0L;
    private final long intervalMs;

    public MrzImageAnalyzer(Context ctx,
                           OcrEngine mlKit,
                           OcrEngine tess,
                           DualOcrRunner.Mode mode,
                           long intervalMs,
                           Listener listener) {
        this.appContext = ctx.getApplicationContext();
        this.mlKitEngine = mlKit;
        this.tessEngine = tess;
        this.mode = mode;
        this.intervalMs = intervalMs;
        this.listener = listener;
        this.aggregator = new MrzBurstAggregator(3, 12);
        this.rectAverager = new RectAverager(6, 0.35f);
    }

    public void setMode(DualOcrRunner.Mode mode) {
        this.mode = mode;
    }

    public void resetBurst() {
        finished.set(false);
        aggregator.reset();
        rectAverager.reset();
    }

    @Override
    public void analyze(@NonNull ImageProxy image) {
        int rotationDeg = image.getImageInfo().getRotationDegrees();
        if (finished.get()) {
            image.close();
            return;
        }
        long now = System.currentTimeMillis();
        if (now - lastTs < intervalMs) {
            image.close();
            return;
        }
        lastTs = now;

        try {
            Bitmap frame = imageProxyToBitmap(image);
            if (frame == null) return;

            if (rotationDeg != 0) {
                Matrix m = new Matrix();
                m.postRotate(rotationDeg);
                frame = Bitmap.createBitmap(frame, 0, 0, frame.getWidth(), frame.getHeight(), m, true);
                rotationDeg = 0;
            }

            Rect detected = MrzAutoDetector.detect(frame);
            if (detected == null) return;

            Rect stable = rectAverager.update(detected, frame.getWidth(), frame.getHeight());
            Bitmap roiBmp = Bitmap.createBitmap(frame, stable.left, stable.top, stable.width(), stable.height());

            DualOcrRunner.RunResult rr = DualOcrRunner.run(appContext, mode, mlKitEngine, tessEngine, roiBmp, rotationDeg);
            if (listener != null) listener.onOcr(rr.ocr, rr.mrz, stable);

            if (rr.mrz != null) {
                MrzResult finalMrz = aggregator.addAndMaybeAggregate(rr.mrz);
                if (finalMrz != null) {
                    finished.set(true);
                    if (listener != null) listener.onFinalMrz(finalMrz, stable);
                }
            }
        } catch (Throwable ignore) {
        } finally {
            image.close();
        }
    }

    private Bitmap imageProxyToBitmap(ImageProxy image) {
        try {
            ImageProxy.PlaneProxy[] planes = image.getPlanes();
            ByteBuffer yBuffer = planes[0].getBuffer();
            ByteBuffer uBuffer = planes[1].getBuffer();
            ByteBuffer vBuffer = planes[2].getBuffer();

            int ySize = yBuffer.remaining();
            int uSize = uBuffer.remaining();
            int vSize = vBuffer.remaining();

            byte[] nv21 = new byte[ySize + uSize + vSize];
            yBuffer.get(nv21, 0, ySize);
            vBuffer.get(nv21, ySize, vSize);
            uBuffer.get(nv21, ySize + vSize, uSize);

            YuvImage yuvImage = new YuvImage(nv21, ImageFormat.NV21, image.getWidth(), image.getHeight(), null);
            try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
                yuvImage.compressToJpeg(new Rect(0, 0, image.getWidth(), image.getHeight()), 90, out);
                byte[] bytes = out.toByteArray();
                return BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
            }
        } catch (Throwable t) {
            return null;
        }
    }
}

File: eMRTDReader/app/src/main/java/com/example/emrtdreader/analyzer/MrzImageAnalyzer.kt
package com.example.emrtdreader.analyzer

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.example.emrtdreader.utils.MrzBurstAggregator
import com.example.emrtdreader.utils.MrzNormalizer
import com.example.emrtdreader.utils.MrzResult
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import java.io.ByteArrayOutputStream
import kotlin.coroutines.resume

class MrzImageAnalyzer(
    private val onFinalMrz: (MrzResult) -> Unit
) : ImageAnalysis.Analyzer {

    private val scope = CoroutineScope(Dispatchers.Default)
    private val aggregator = MrzBurstAggregator()
    private var finished = false

    private var lastAnalysisTime = 0L
    private val analysisIntervalMs = 300L

    @androidx.camera.core.ExperimentalGetImage
    override fun analyze(image: ImageProxy) {
        if (finished) {
            image.close()
            return
        }

        val now = System.currentTimeMillis()
        if (now - lastAnalysisTime < analysisIntervalMs) {
            image.close()
            return
        }
        lastAnalysisTime = now

        scope.launch {
            try {
                processImage(image)
            } finally {
                image.close()
            }
        }
    }

    private suspend fun processImage(image: ImageProxy) {
        val bitmap = imageProxyToBitmap(image)
        val mrzBitmap = cropMrzRegion(bitmap)

        val ocrText = runOcr(mrzBitmap)
        val normalizedResult = MrzNormalizer.normalize(ocrText)
        
        val aggregated = aggregator.aggregate(normalizedResult)
        if (aggregated != null && aggregated.confidence >= 3) {
            finished = true
            onFinalMrz(aggregated.result)
            aggregator.reset()
        }
    }

    private suspend fun runOcr(bitmap: Bitmap): String =
        suspendCancellableCoroutine { cont ->
            val image = InputImage.fromBitmap(bitmap, 0)
            TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
                .process(image)
                .addOnSuccessListener { if (cont.isActive) cont.resume(it.text) }
                .addOnFailureListener { if (cont.isActive) cont.resume("") }
        }

    private fun cropMrzRegion(bitmap: Bitmap): Bitmap {
        val h = bitmap.height
        val w = bitmap.width
        return Bitmap.createBitmap(bitmap, 0, (h * 0.65).toInt(), w, (h * 0.35).toInt())
    }

    private fun imageProxyToBitmap(image: ImageProxy): Bitmap {
        val yBuffer = image.planes[0].buffer
        val uBuffer = image.planes[1].buffer
        val vBuffer = image.planes[2].buffer

        val ySize = yBuffer.remaining()
        val uSize = uBuffer.remaining()
        val vSize = vBuffer.remaining()

        val nv21 = ByteArray(ySize + uSize + vSize)

        yBuffer.get(nv21, 0, ySize)
        vBuffer.get(nv21, ySize, vSize)
        uBuffer.get(nv21, ySize + vSize, uSize)

        val yuvImage = YuvImage(nv21, ImageFormat.NV21, image.width, image.height, null)
        val out = ByteArrayOutputStream()
        yuvImage.compressToJpeg(Rect(0, 0, image.width, image.height), 90, out)
        val jpegBytes = out.toByteArray()

        return BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.size)
    }
}